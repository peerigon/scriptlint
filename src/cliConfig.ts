import commander from "commander";
// eslint-disable-next-line @typescript-eslint/no-var-requires
const { version } = require("../package.json");

type CliConfig = {
	packageFile?: string;
	fix?: boolean;
	strict?: boolean;
	json?: boolean;
	config?: boolean;
};

export default (argv: Array<string>): CliConfig => {
	const program = new commander.Command();
	let packageFile;

	program
		.version(`${version}`)
		.arguments("[packageFile]")
		.action((arg: string) => {
			packageFile = arg;
		})
		.option("-s, --strict", "strict mode")
		.option("-j, --json", "JSON output")
		.option("-c, --config", "inspect the config")
		.option("-f, --fix", "autofixing");

	program.parse(argv);

	// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
	const cliConfig: CliConfig = { packageFile: packageFile ?? process.cwd() };

	if (program.fix !== undefined) {
		cliConfig.fix = program.fix;
	}

	if (program.strict !== undefined) {
		cliConfig.strict = program.strict;
	}

	if (program.json !== undefined) {
		cliConfig.json = program.json;
	}

	if (program.config !== undefined) {
		cliConfig.config = program.config;
	}

	return cliConfig;
};
