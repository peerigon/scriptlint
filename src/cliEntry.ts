import cliRun from "./cli";

cliRun(process.argv);
