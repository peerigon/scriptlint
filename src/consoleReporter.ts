import chalk from "chalk";
import { Values, MessageBuffer, MessageType } from "./types";
import { makeMessage } from "./utils";

const PREFIX = "⧙։⧘";
let stashed: MessageBuffer = [];

const stash = (message: string, type: MessageType): void => {
	stashed.push({
		message,
		type
	});
};

export const warning = (template: string, values?: Values): void => {
	const message = makeMessage(template, values);

	stash(message, "warning");
};

export const success = (template: string, values?: Values): void => {
	const message = makeMessage(template, values);

	stash(message, "success");
};

export const dump = (jsonOutput: boolean): number => {
	const problemCount = stashed.length;

	if (jsonOutput) {
		// eslint-disable-next-line no-console
		console.log(JSON.stringify(stashed, null, "\t"));
	} else {
		stashed.forEach(({ message, type }) => {
			print(type, message);
		});
	}

	stashed = [];

	return problemCount;
};

export const error = (message: string): void => {
	print("error", message);
};

const print = (type: MessageType, message: string) => {
	switch (type) {
	case "error": {
		// eslint-disable-next-line no-console
		console.log(chalk.bold.red(`${PREFIX} [error] ${message}`));
		break;
	}

	case "warning": {
		const notUnderlined = chalk.yellow.bold(`${PREFIX} [${type}]`);
		const underlined = chalk.yellow.underline(message);

		// eslint-disable-next-line no-console
		console.log(`${notUnderlined} ${underlined}`);
		break;
	}
	case "success": {
		// eslint-disable-next-line no-console
		console.log(chalk.bold.gray(`${PREFIX} [✔️] ${message}`));
		break;
	}
	}
};
