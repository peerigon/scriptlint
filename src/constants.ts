import {Config} from "./types";

export const DEFAULT_CONFIG: Config = {
	strict: false,
	fix: false,
	json: true,
	config: false,
	packageFile: undefined,
	rules: {},
	customRules: [],
	ignoreScripts: [],
};

export const PROJECT_NAME = "scriptlint";

export const NAME_REGEX = /^[\d:A-Za-z]+$/;

export const NAMESPACES = [
	"build",
	"dev",
	"format",
	"other",
	"report",
	"setup",
	"start",
	"test"
];

export const DEFAULT_NPM_HOOKS = [
	"install",
	"postinstall",
	"postpack",
	"postpublish",
	"postrestart",
	"postshrinkwrap",
	"poststart",
	"poststop",
	"posttest",
	"postuninstall",
	"postversion",
	"preinstall",
	"precommit",
	"prepack",
	"prepare",
	"prepublish",
	"prepublishOnly",
	"prerestart",
	"preshrinkwrap",
	"prestart",
	"prestop",
	"pretest",
	"preuninstall",
	"preversion",
	"publish",
	"release",
	"restart",
	"shrinkwrap",
	"stop",
	"uninstall",
	"version"
];
