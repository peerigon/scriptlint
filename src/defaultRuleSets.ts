import defaultRules from "./rules";
import { Rule } from "./types";

export const optionalRules = ["natural-order"];

const strict = defaultRules
	.map((r: Rule) => r.name)
	.filter((r) => !optionalRules.includes(r));

const ruleSets = {
	default: [
		"mandatory-test",
		"mandatory-start",
		"mandatory-dev",
		"no-default-test",
	],
	strict,
	optionalRules,
};

export default ruleSets;
