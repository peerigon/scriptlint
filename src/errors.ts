export class ConfigError extends Error {
	constructor(message: string) {
		super(message);
		Object.setPrototypeOf(this, ConfigError.prototype);
	}
}

export class ValidationFunctionInvalidError extends Error {
	constructor(name: string) {
		super(`Rule validation function is not a function (${name})`);
		Object.setPrototypeOf(this, ValidationFunctionInvalidError.prototype);
	}
}

export class PackageFileNotFoundError extends Error {
	constructor(path: string) {
		super("package.json could not be found at this location: " + path);
		Object.setPrototypeOf(this, PackageFileNotFoundError.prototype);
	}
}
