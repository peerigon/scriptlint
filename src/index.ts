/*
路---------路
| {scr:pt |
|   l:nt} |
|         |
|         |
路---------路
*/

import module from "./module";

export = module;
