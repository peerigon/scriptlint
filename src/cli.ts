#!/usr/bin/env node
import loadUserConfig from "./userConfig";
import loadCliConfig from "./cliConfig";
import { makePackageFilePath } from "./utils";
import userPackageScriptContext from "./userPackageScripts";
import { loadRulesFromRuleConfig } from "./loadRules";
import execute from "./execute";
import { success, warning, dump, error } from "./consoleReporter";
import {
	DEFAULT_CONFIG,
	PROCESS_EXIT_ERROR,
	PROCESS_EXIT_OK
} from "./constants";

/* istanbul ignore next */
const processExit = (code: number) => {
	if (typeof process.env.JEST_WORKER_ID === "undefined") {
		// eslint-disable-next-line no-process-exit
		process.exit(code);
	}
};

export default (argv: Array<string>) => {
	try {
		/**
		 * config assembly
		 */

		const userConfig = loadUserConfig();
		const cliConfig = loadCliConfig(argv);
		const config = {
			...DEFAULT_CONFIG,
			...{ json: false },
			...userConfig,
			...cliConfig
		};

		// output the config (--config) but only if we don't want JSON output on the CLI
		if (!config.json && config.config) {
			const json = JSON.stringify(config, null, 2);

			// eslint-disable-next-line no-console
			console.log(`\n\n${json}\n\n`);
		}

		/**
		 * package.json
		 */

		const {
			writePackageScripts,
			readPackageScripts
		} = userPackageScriptContext(
			makePackageFilePath(config.packageFile ?? process.cwd())
		);

		const scripts = readPackageScripts(config.ignoreScripts);

		/**
		 * rule loading
		 */

		const rules = loadRulesFromRuleConfig(
			config.strict,
			config.rules,
			config.customRules
		);

		/**
		 * and go!
		 */
		const [issues, fixedScripts, issuesFixed] = execute(
			rules,
			scripts,
			warning,
			config.fix
		);

		if (config.fix && issuesFixed > 0) {
			writePackageScripts(fixedScripts);
			success(`Fixed ${issuesFixed} issue${issuesFixed > 1 ? "s" : ""}!`);
		}

		if (issues.length - issuesFixed === 0) {
			success("âœ¨  All good");

			dump(config.json);

			processExit(PROCESS_EXIT_OK);
		}

		dump(config.json);
		processExit(PROCESS_EXIT_ERROR);
	} catch (err) {
		error(err);
	}
};
