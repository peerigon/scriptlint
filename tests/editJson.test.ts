import EditJson from "../src/editJson";
import { PackageFileNotFoundError } from "../src/errors";
// eslint-disable-next-line @typescript-eslint/no-var-requires
const detectIndent = require("detect-indent");

jest.mock("fs");

describe("editJson.ts", () => {
	describe("it throws on file not found", () => {
		expect(() => new EditJson("foo/bar/baz")).toThrow(
			PackageFileNotFoundError
		);
	});

	it("has a default scripts section", () => {
		const fileWithoutScripts = new EditJson(
			"real/existing/path/package-without-scripts.json"
		);

		expect(fileWithoutScripts.get().scripts).toEqual({});
	});

	const file = new EditJson("real/existing/path/package.json");

	it("reads files", () => {
		expect(file).toMatchSnapshot();
	});
	test("get()", () => {
		expect(file.get()).toEqual({ scripts: { foo: "bar" } });
	});

	test("set()", () => {
		expect(file.set("scripts", { foo: "bar" }).get()).toEqual({
			scripts: { foo: "bar" }
		});
	});

	it("should preserve indentation", () => {
		expect(file.indent).toEqual({ amount: 2, indent: "  ", type: "space" });
	});

	const saved = file.save();

	test("save()", () => {
		expect((saved as any).written).toBe(true);
	});

	it("respects indentation 1", () => {
		expect(detectIndent((saved as any).content).type).toEqual("space");
	});

	it("respects indentation 2", () => {
		const file2 = new EditJson("real/existing/path/package-with-tabs.json");

		file2.set("scripts", { foo: "bar" });
		const saved2 = file2.save();

		expect(detectIndent((saved2 as any).content).type).toEqual("tab");
	});
});

export {};
