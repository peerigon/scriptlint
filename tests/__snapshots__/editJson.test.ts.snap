// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editJson.ts reads files 1`] = `
default_1 {
  "fileContents": "{
  "scripts": {
    "foo": "bar"
  }
}",
  "indent": {
    "amount": 2,
    "indent": "  ",
    "type": "space",
  },
  "package": {
    "scripts": {
      "foo": "bar",
    },
  },
  "path": "real/existing/path/package.json",
}
`;
